components/controls/Tooltip.tsx

```diff
@@ -252,8 +286,85 @@ export class TooltipInner extends React.Component<TooltipProps, State> {
     this.handleMouseLeave();
   };
 
-  render() {
+  needsFlipping = (leftFix: number, topFix: number) => {
+    // We can live with a tooltip that's slightly positioned over the toggle
+    // node. Only trigger if it really starts overlapping, as the re-positioning
+    // is quite expensive, needing 2 re-renders.
+    const threshold = this.context.theme.rawSizes.grid;
+    switch (this.getPlacement()) {
+      case 'left':
+      case 'right':
+        return Math.abs(leftFix) > threshold;
+      case 'top':
+      case 'bottom':
+        return Math.abs(topFix) > threshold;
+    }
+    return false;
+  };
+
+  renderActual = ({ leftFix = 0, topFix = 0 }) => {
+    if (
+      !this.state.flipped &&
+      (leftFix !== 0 || topFix !== 0) &&
+      this.needsFlipping(leftFix, topFix)
+    ) {
+      // Update state in a render function... Not a good idea, but we need to
+      // render in order to know if we need to flip... To prevent React from
+      // complaining, we update the state using a setTimeout() call.
+      setTimeout(() => {
+        this.setState(
+          ({ placement = 'bottom' }) => ({
+            flipped: true,
+            // Set height to undefined to force ScreenPositionFixer to
+            // re-compute our positioning.
+            height: undefined,
+            placement: FLIP_MAP[placement]
+          }),
+          () => {
+            if (this.state.visible) {
+              // Force a re-positioning, as "only" updating the state doesn't
+              // recompute the position, only re-renders with the previous
+              // position (which is no longer correct).
+              this.positionTooltip();
```

Comment:
This will also call a `setState`, so the there will be two renders triggered from this ? for the first `setState` that updates the `flipped` and then the one for the `positionTooltip`, is it the two renders you are talking about in `needsFlipping` ?

Reply:
React can't batch those `setState` because you use the callback to trigger `positionTooltip` (since it requires the placement to be up to date in the state.
But could pass the new placement directly to `positionTooltip` to avoid using the callback and maybe that would allow react to batch those `setState` and only render once ?

Reply:
Good catch! No, this wouldn't trigger 2 re-renders, but (potentially) **4** (because of `ScreenPositionFixer`)!

I will call `setState()` only once 👍 .

Reply:
No need to call it only once, it might be difficult to put in place ? Just allowing to pass the new placement to positionTooltip so it doesn't have to rely on the state could be enough, then react would be able to batch those state change and trigger only one render.

Reply:
Hmm, I realize it won't work. We need to explicitly set `height` to undefined, which triggers a re-render of `ScreenPositionFixer`, which is then ready to re-compute the positioning once more.

For a non-flipping Tooltip, the flow looks like this (not counting the initial rendering of the toggle node itself):

![Screenshot 2019-08-23 at 14 03 30](https://user-images.githubusercontent.com/45544358/63591447-1db72000-c5af-11e9-91f2-d624df4dcab7.png)

For a flipping Tooltip, it looks like this:

![Screenshot 2019-08-23 at 14 04 06](https://user-images.githubusercontent.com/45544358/63591483-39bac180-c5af-11e9-9267-4732999cb8d3.png)

So, by default, **without** my fix, a Tooltip renders twice when shown, and twice when hidden (total of 4 times).
**With** my fix, a Tooltip renders four times when shown, and twice when hidden (a total of 6 times).

I think I might be able to cut this to three times, at a slight cost of readability. But is the ROI high enough, considering:

1. This will happen pretty rarely
2. We're "only" re-rendering a Tooltip, not a whole section of the application. Is winning a single render call, in rare situations, really worth that extra effort and "cost" in maintainability?

Reply:
Ok thanks for the details, indeed I don't think it's worth investing more into this, let's keep your current implementation! :+1: